runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -11963,3 +11986,295 @@ J9::Z::TreeEvaluator::tabortEvaluator(TR::Node * node, TR::CodeGenerator * cg)
 #endif
    return NULL;
    }
+
+TR::Register *
+J9::Z::TreeEvaluator::ardbarEvaluator(TR::Node * node, TR::CodeGenerator * cg)
+   {
+   bool dynLitPoolLoad = false;
+   TR::Register* resultReg = TR::TreeEvaluator::checkAndAllocateReferenceRegister(node, cg, dynLitPoolLoad);
+
+   if (TR::Compiler->om.shouldReplaceGuardedLoadWithSoftwareReadBarrier())
+      {
+      resultReg = J9::Z::TreeEvaluator::generateSoftwareReadBarrier(node, cg, resultReg, NULL,
+                                                                    cg->comp()->useCompressedPointers(),
+                                                                    true);
+      }
+   else
+      {
+      TR::MemoryReference* loadMemRef = generateS390MemoryReference(node, cg);
+      TR::TreeEvaluator::checkAndSetMemRefDataSnippetRelocationType(node, cg, loadMemRef);
+      TR::InstOpCode::Mnemonic loadOp = cg->comp()->useCompressedPointers() ? TR::InstOpCode::LLGFSG : TR::InstOpCode::LGG;
+      generateRXInstruction(cg, loadOp, node, resultReg, loadMemRef);
+      node->setRegister(resultReg);
+      }
```

Comment:
Shouldn't the `else` path here be an `else if` and the `else` path be calling the base `ardbarEvaluator` implementation at the common level? That is, after your changes are synchronized with the ones over in #4743.

In fact, shouldn't this look more like the case above in `VMinlineCompareAndSwap`? i.e.

```
   if (TR::Compiler->om.shouldGenerateReadBarriersForFieldLoads())
      {
      if (TR::Compiler->om.shouldReplaceGuardedLoadWithSoftwareReadBarrier())
         {
         ...
         }
      else
         {
         ...
         }
      }
   else
      {
      // Call base evaluator
      }
```

Reply:
Yes. I was considering to do this when #4743 is ready to merge. We'll conflict in this evaluator anyway.

Reply:
We should go ahead and mold the code into the above format in preparation. I think this will be merged before #4743.

Reply:
This is addressed as part of 32009de1e6.